### ISSUE TITLE
Handling `self`'s type inside decorating functions.

### ISSUE THREAD

**Issue by jbkoh** [microsoft/pyright#6472]
**Describe the bug**
When `self` argument is explicitly defined in a decorating/wrapping function and then used it by a class method, the type checker considers it to be a required positional argument even if it's used from an instance. Thus, I get 
> [Pyright reportGeneralTypeIssues] [E] Argument missing for parameter "self"

**Code or Screenshots**

```python

from typing import Callable
from functools import wraps

def print_something(f: Callable):
    @wraps(f)
    def wrapper(self, *args, **kwargs):
        print('hello')
        return f(self, *args, **kwargs)
    return wrapper


class TestClass(object):

    @print_something
    def test(self):
        pass


instance = TestClass()
instance.test() # [Pyright reportGeneralTypeIssues] [E] Argument missing for parameter "self"
```

The last line causes an error as in the comment.

**VS Code extension or command-line**
I'm using coc-pyright in nvim.

**Comment by erictraut** [comment]
You're missing some required type information in your decorator. Here's the proper way to annotate it:

```python
P = ParamSpec("P")
R = TypeVar("R")
S = TypeVar("S")


def print_something(
    f: Callable[Concatenate[S, P], R]
) -> Callable[Concatenate[S, P], R]:
    @wraps(f)
    def wrapper(self: S, *args: P.args, **kwargs: P.kwargs):
        print("hello")
        return f(self, *args, **kwargs)

    return wrapper
```

Or if you're using Python 3.12, you can use the new PEP 695 notation:

```python
def print_something[**P, R, S](
    f: Callable[Concatenate[S, P], R]
) -> Callable[Concatenate[S, P], R]:
    @wraps(f)
    def wrapper(self: S, *args: P.args, **kwargs: P.kwargs):
        print("hello")
        return f(self, *args, **kwargs)

    return wrapper
```
